src/battle_script_commands.c

```diff
@@ -3801,6 +3801,53 @@ void SetMoveEffect(bool32 primary, u32 certain)
                     gBattlescriptCurrInstr = BattleScript_SpikesActivates;
                 }
                 break;
+            case MOVE_EFFECT_TRIPLE_ARROWS:
+                {
+                    u8 randomLowerDefenseChance;
+                    u8 randomFlinchChance;
+
+                    if (GetBattlerAbility(gBattlerAttacker) == ABILITY_SERENE_GRACE)
+                    {
+                        randomLowerDefenseChance = RandomPercentage(RNG_TRIPLE_ARROWS_DEFENSE_DOWN, 100);
+                        randomFlinchChance = RandomPercentage(RNG_TRIPLE_ARROWS_FLINCH, 60);
+                    }
+                    else
+                    {
+                        randomLowerDefenseChance = RandomPercentage(RNG_TRIPLE_ARROWS_DEFENSE_DOWN, 50);
+                        randomFlinchChance = RandomPercentage(RNG_TRIPLE_ARROWS_FLINCH, 30);
+                    }
+
+                    if (randomLowerDefenseChance && randomFlinchChance)
+                    {
+                        BattleScriptPush(gBattlescriptCurrInstr + 1);
+                        gBattlescriptCurrInstr = BattleScript_ReduceDefenseAndFlinch;
+                    }
+                    else
+                    {
+                        if (randomLowerDefenseChance)
+                        {
+                            BattleScriptPush(gBattlescriptCurrInstr + 1);
+                            gBattlescriptCurrInstr = BattleScript_DefDown;
+                        }
+                        else if (randomFlinchChance)
+                        {
+                            if (battlerAbility == ABILITY_INNER_FOCUS)
+                            {
+                                gLastUsedAbility = ABILITY_INNER_FOCUS;
+                                gBattlerAbility = gEffectBattler;
+                                RecordAbilityBattle(gEffectBattler, ABILITY_INNER_FOCUS);
+                                gBattlescriptCurrInstr = BattleScript_FlinchPrevention;
+                            }
+                            else
+                            {
+                                if (GetBattlerTurnOrderNum(gEffectBattler) > gCurrentTurnActionNumber)
+                                    gBattleMons[gEffectBattler].status2 |= sStatusFlagsForMoveEffects[MOVE_EFFECT_FLINCH];
+                                gBattlescriptCurrInstr++;
+                            }
+                        }
+                    }
+                }
+                break;
```

Comment:
This whole section is still incredibly hacky with a lot of duplicate code. Even if it works, it would be best if cleaned up.

Reply:
I didn't keep it like that because it worked, but because I couldn't come up with something better *and* that worked.

I've been staring really hard and scratching my peanut sized brain some more, and I think this is a nice improvement and it still seems to work. Probably the farthest I can personally go.

https://diffy.org/diff/9f85000bb97c8

Reply:
I don't think you're giving yourself enough credit 🙂
Take it slow, and take a break if you need to. I'm sure after enough time you'll be able to figure it out! Hacky code can often lead to unwanted bugs down the line. It's better to take it slow and write something that if modified in the future will stay bug-free.

Reply:
I understand that, that's why I haven't resolved the conversation.

Still, I can't do more than I did here, so we're at an impasse.

Reply:
Just do
```
u8 randomLowerDefenseChance = RandomPercentage(RNG_TRIPLE_ARROWS_DEFENSE_DOWN, GetMoveSecondaryEffectChance(gBattlerAttacker, 50));
u8 randomFlinchChance = RandomPercentage(RNG_TRIPLE_ARROWS_FLINCH, GetMoveSecondaryEffectChance(gBattlerAttacker, 30));

if (randomFlinchChance
&& battlerAbility != ABILITY_INNER_FOCUS
&& GetBattlerTurnOrderNum(gEffectBattler) > gCurrentTurnActionNumber)
{
    gBattleMons[gEffectBattler].status2 |= sStatusFlagsForMoveEffects[MOVE_EFFECT_FLINCH];
}

if (randomLowerDefenseChance)
{
    BattleScriptPush(gBattlescriptCurrInstr + 1);
    gBattlescriptCurrInstr = BattleScript_DefDown;
}
else
    gBattlescriptCurrInstr++;
```

If your tests fail because of this change, might be worth reworking those.